### ISSUE TITLE
The future: Scalaz, Cats, both, or neither?

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#390]
## Ground rules

Scalaz and Cats provide a lot of overlapping functionality that is useful to our library.  Each has a both a devoted following and strong detractors.  We are not here to discuss history or politics, but to transparently decide the way forward for this project.
## The present

Currently, we are a unabashedly a Scalaz library.  We get this transitive dependency from scalaz-stream, so we might as well use it.  
## The looming change

Soon, [scalaz-stream will be decoupled](https://github.com/scalaz/scalaz-stream#issues/321).  The approach taken there is to not depend on either Scalaz or Cats, and redefine its own foundational type classes (e.g., Monad) as necessary.

Once we depend on this version of scalaz-stream, tentatively to be branded 0.9, the Scalaz decision is no longer made for us.
### Decouple core from either library

Pros:
- Follows streams' philosophy that "The ecosystem should not pick a winner." 
- Imposes no new dependencies on us beyond streams.
- Does not bifurcate our community.

Cons:
- Forces us to settle for inferior types (e.g., `scala.Either` vs. `scalaz.\/`) or reinvent them.
- Other similarly principled libraries (e.g., doobie, remotely) will need similar reinventions, and without using a library, provide a worse interop experience.
- Instances could be supplied by http4s-scalaz and/or http4s-cats integration modules, but would no longer conveniently appear in companion objects.

Short term impact:
- We should move all direct Scalaz dependencies to a support module or delete them.
### Build on Scalaz

Pros:
- No additional breaking changes
- Matches current state of doobie and remotely.  This is no guarantee those projects don't change.

Cons:
- Gives second-class status to Cats projects.
- Scalaz is already discussing an incompatible version 8.  That's basically a whole new library.

Short term impact:
- None
### Build on Cats

Pros:
- Cats has a rapidly growing community.
- Our existing maintainers contribute more to Cats than Scalaz.

Cons:
- Less mature than Scalaz, further pushing out our timeline for a long-term support API.
- Gives second-class status to Scalaz projects, which are presently more numerous.

Short term impact:
- Keep building on Scalaz, as long as the concept is or is expected to appear in Cats.  Some API will change, but the concepts are the same.
- <ins>Consider migrating instance layout to match Cats conventions</ins>.
### Cross-build

Pros:
- http4s is positioned well if either or both libraries continue on strong.

Cons:
- Maintaining cross-build stretches thin development team even thinner.
- Some who have tried cross-building advise that this way madness lies.
- Depending how we do it, restricts us to the intersection of functionality supported by Scalaz and Cats.

Short term impact:
- None

**Comment by bryce-anderson** [comment]
Cross builds are a non-starter for me: I just don't want to deal with it. That is not the fun part of coding (for me) and I'm not (currently) paid for this. If this can be done without me ever having to care, cool, but that seems unlikely.

Most projects are sitting on the fence watching the scala FP landscape smolder. I don't have any reason to, so I'm going to suggest we make sure Cats is going in a sane direction and, if it is, go with them when agnostic streams lands. Maybe other projects will follow, maybe they wont. I'm willing to bet on the welcoming community because that is where I would feel more comfortable contributing and thus learning, and that is really the only reason I'm doing this.

**Comment by rossabaker** [comment]
I have looked to scalaz-stream as the precedent setter on this subject, but we should not overlook Argonaut/Circe, where the answer was a fork.  I think a fork is a more plausible form of "cross-build without ever having to care".  Now both Scalaz and Cats now have a commendable JSON library.  I don't see this as a tragic outcome, though watching divergence and cross-pollination in the coming months will be fascinating. 

I tried to be the neutral moderator in the issue description, and now speak personally.  I respect the "ecosystem should not pick a winner" philosophy of streams, but I find the cost of reinventing twice invented concepts too high.  My first preference is to participate in a cohesive FP ecosystem, and my second preference is for Cats to be the basis.

**Comment by mdedetrich** [comment]
I personally prefer Cats, and the preference is due to a technical one. Cats puts an emphasis on modularity and stability (API/Binary stability, massive emphasis on testing). I have been bitten many times in the past due to Scalaz's monolithic design (coupled with its its attempt to reinvent the scala stdlib). At least for the medium/long term, it seems a better decision as you would strictly only need to pull in the parts of cats that you need (which would probably be whatever a cats-stream/cats-task would pull in)

In regards to pathways, I don't see the merit in supporting both Cats/Scalaz. Its a huge amount of work, and may be setting the wrong precedent. Http4s should be using the library which best aligns with its goals of stability, and Cats is follows this a lot harder than Scalaz.

For now, I would also decouple JSON from this discussion, I think its a separate issue, mainly for the reason we can always add in JSON support for library X through some external modules (i.e. http4s-Argonaut). Such modules can bring in whatever they want, but stuff like stream/task is so central to http4s, that it really needs to use something that is more stable/modular

**Comment by ShengC** [comment]
Or, 4th option - mostly depend on either `Cats` or `Scalaz`, while keeping some of the types agnostic, i.e., maintain by ourselves. Take `Task` for example, it is used everywhere across our code base, having it agnostic gives us the flexibility to tailor it to our own use case in the future.

**Comment by rossabaker** [comment]
Are you proposing we maintain our own Task?

Historical note: we considered, very early on, abstracting over Task in the services.  One reason we didn't was that it created an awful lot of noise when the other use case anyone had was `IO`, which is mainly of utility in a Servlet 2.5 container, which nobody is going to use these days.  A second reason was that so many of the interesting things built into scalaz-stream are already fixed on Task.  

We could create and maintain our own Task, but that won't be the Task that doobie and remotely and others use.  If the new scalaz-stream is less Task-centric, we might reopen that discussion one more time, but it's a whole lot of effort unless somebody has a really compelling reason that they don't want to use streams' Task.

**Comment by ShengC** [comment]
Let's possibly revisit this when the agnostic stream actually ships out. No
matter we choose `cats` or `scalaz` in the future, we would end up with 2
different `Task` unless we decide to replace `EntityBody` with another data
structure. I think if we maintain our own `Task`, at least we would have
full control to minimize the gap between the two with some properly
maintained implicits. Of course this is debatable and may introduce more
complications, but at least something we can think of.

**Comment by mdedetrich** [comment]
Would also like to say that I disagree quite a bit with cross building from a purely tech POV. Once you start using stuff like scalaz-stream (or its cats equivalent) in a non trivial way, you get a lot of subtleties which can differ between implementations, which can make things that much harder in both supporting and documenting both implementations.

My personal preference is for hanging out with Scalaz currently (maybe make a 1.0 that uses Scalaz), and then migrate over to Cats when we are ready to do so.

**Comment by rossabaker** [comment]
I agree that no final decision must be made now.  I wanted to open this now to:
* Inform short term discussions of building on existing types vs. custom ones (e.g., #389)
* Give fair warning to those who prefer Scalaz that a long term shift to Cats is under consideration.

I think we should continue to embrace Scalaz in the present as it's already there, is good technology, and really not that terrible to port to Cats.  Once the decoupled stream library is released, we can look at our community and the wider FP-in-Scala community and make a more informed decision on this issue for the long haul.

**Comment by dchenbecker** [comment]
:+1: to that last comment by @rossabaker. Given my experience just supporting a cross-build for different *Scala* versions back in Lift-land I can imagine the pain and suffering you would endure trying to support two similar-but-different libraries. Generally, I think that as Cats progresses there will be fewer pain points in a port but unless and until it's a low-cost maneuver I don't see a big gain. I suppose the near future will give us a better idea of direction for both libraries and hopefully confirm some of the supposition around how Cats will work out. I think that Bryce's observation about the community around Cats is one that, if it bears out, is a strong indicator of (finally) a positive outcome from the whole fiasco.

Regardless, thanks for opening this discussion up!

**Comment by rossabaker** [comment]
@djspiewak has gifted the world with [shims](https://github.com/djspiewak/shims).  Kneejerk and probably inadequate analysis:

1. Neatly solves our needs for typeclasses without alienating part of our community.
2. We still have to be opinionated on, or reinvent, data types (e.g., DecodeResult).  This may be a dealbreaker, even though we seem very much a target audience for a library like shims.
3. Does not at insulate us from upstream libraries' choice of data types inharmonious with ours (e.g., fs2 and `scala.Either`).

**Comment by djspiewak** [comment]
@rossabaker I think your "kneejerk" analysis is pretty accurate.  With that said, http4s was one of the projects I was specifically thinking about when @alissapajer and I decided to do shims, so I definitely lend credence to your concerns.  :-)

Here's the bottom line: I don't want to replicate scalaz (or cats).  We already have two such frameworks; we really don't need (or want) a third.  Paul O'Keeffe pointed out on twitter today that the whole situation is very reminiscent of the logger framework proliferation on the JVM over the last decade or so, and I think that is (unfortunately) an apt analogy.  All I want is to do the work, *once*, that every framework author who wants to support cats and scalaz together would otherwise have to do.

There's a lot of wiggle room in that mission statement though.  Like… maybe we *should* insulate you from `Xor` vs `\/` (it would certainly make fs2's life easier!).  I'm very open to thoughts and opinions on this, or more selfishly: pull requests.  :-)  Please by all means expand shims to to fill the needs you see!  I mean, I still think that `State` is pretty much out of scope, and something like `DecodeResult` is *definitely* out of scope, but the library as it stands is just the bare bones necessary to get `Emm` compiling.  It's far from complete.

**Comment by ashugupt** [comment]
@rossabaker Has a consensus been reached over this? In my new project I intend to go with cats(nothing against scalaz) and would like to use HTTP4S with fs2/stream going forward in lieu of akka-http simply due to the sheer amount of implicit def magic it comes with. If in some near future timeline, http4s has plans to align with cats (or be agnostic) I would love to onboard.

**Comment by j-keck** [comment]
Am 30. Juni 2016 05:44:53 MESZ, schrieb Ashu Gupta <notifications@github.com>:
>@rossabaker Has a consensus been reached over this? In my new project I
>intend to go with cats(nothing against scalaz) and would like to use
>HTTP4S with fs2/stream going forward in lieu of akka-http simply due to
>the sheer amount of implicit def magic it comes with. If in some near
>future timeline, http4s has plans to align with cats (or be agnostic) I
>would love to onboard.
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/http4s/http4s/issues/390#issuecomment-229552400

-- 
gesendet vom phone

**Comment by rossabaker** [comment]
I am definitely interested in full support for cats.  Whether that's by making http4s-core depend on fs2 and cats, building on shims, or an http4s-cats module that sits on top of an agnostic core, is an open question.  I think there are pros and cons to all three approaches that might best be decided by trying one and seeing how it feels.

Note that it's possible to use cats on an http4s project today.  If you use http4s-circe, you'll have both cats and scalaz on the classpath.  It's not ideal to have two libraries that do almost the same thing, but I've done this in production, and it didn't hurt.

**Comment by rossabaker** [comment]
Doobie has an [interesting approach](https://github.com/tpolecat/doobie/pull/304).

I think this might work for cats and scalaz.  I think streaming is much more intrinsic to http4s than it is to doobie, so I'm not as certain this works for supporting scalaz-stream and fs2.  I think the approach there might need to be to cut over to fs2, and provide a scalaz-stream/scalaz.concurrent.Task support module.

**Comment by rossabaker** [comment]
Not only does it support both category libraries, it supports both streaming libraries.  It also is transparent: no shim types in the public API, no additional imports.  The pain is entirely the library's.  And it does hurt.

Stacks that could be supported natively with this:

1) scalaz + scalaz-stream
2) scalaz + fs2
3) cats + scalaz-stream
4) cats + fs2

Among these, 1 and 4 are compelling to me.  2 could be compelling to the community, and comes along free with the work in 1 and 4.  3 is the unwanted quadrant, I suspect.

Because scalaz+fs2 is more compelling than cats+scalaz-stream, I'm attempting the fs2 port first.  See the topic/yax branch as I hack and whack.  I am not yet committed to this approach, but would love feedback.  In the worst case, the yax-generated source can seed an eventual move to fs2.

**Comment by tpolecat** [comment]
@djspiewak says "dear god", miles calls it "horrible", dibblego says "haha *eyeroll*" ... this is supremely entertaining if nothing else.

**Comment by bryce-anderson** [comment]
A preprocessor feels like a good way to suppress the contributor count: I don't think it's something I'd be interested in fooling with.

I'd rather support fewer FP libraries.

**Comment by kryptt** [comment]
I agree with @bryce-anderson. I hadn't consciously thought about shying away potential contributors, but I do remember feeling an overbearing pain as I read through the yaxed doobie...

**Comment by tpolecat** [comment]
"feeling an overbearing pain" ... 👍

**Comment by rossabaker** [comment]
I fear that we've entered that transitional period where it's painful to not support both to some degree:

1. We are getting increasing feedback from potential users who are moving forward with cats and fs2.  The status quo serves a shrinking portion of the Scala FP community, which is also bad for future contribution.

2. I am on a very large scalaz-stream code base at a shop that's enthusiastic about http4s.  If http4s drops that support, I will have to maintain an internal fork for a substantial time.

Possible solutions:

1. Maintain two branches in parallel.  Labor intensive.  Feels like a more amiable emacs vs. xemacs, which is still not an endorsement.
2. Yax.  Standing here waist deep in the yax poop, my biggest complaint is that the source tree is no longer Real Scala™, which is my analysis for why it's so poorly reviewed, even by the two people who have done it.
3. "Yax 2.0": no directives, so the main source is still Real Scala™.  It would rewrite imports and import a few syntax classes so `fs2.Task.run` == `fs2.Task.unsafeRun`.  A few things (mostly: Process1) would need to be maintained and referenced in unrewritten source, but the end result is a less burdensome variation of two branches.  Still magical and raises the bar for maintainers, if not contributors.
4. Status quo, but provide -cats and -fs2 support modules.  I've heard from one user that they don't want both category libraries on the classpath and wouldn't use it.
5. Move to cats and fs2, but provide a -scalaz-stream support module.  I've heard from one scalaz-stream user that fs2 isn't production ready yet and he wouldn't use it.
6. Parameterize both the stream and the effect.  Works for scalaz-stream | fs2, but not scalaz | cats.  Application code remains clean, but the internal code drowns in type parameters, and raises the bar to understanding compiler errors.
7. Shims.  I still think it works for scalaz | cats, but not scalaz-stream | fs2.

**Comment by djspiewak** [comment]
Shims could be made to work for scalaz-stream/fs2 to a somewhat limited extent, but it runs into the "`Either` problem" pretty quickly.  Also, I'm dubious that I would be able to uniformly shim a sufficient swath of the library as to meet http4s's needs.

**Comment by rossabaker** [comment]
With Scala 2.12 on the horizon, the Either Problem morphs into the Stuck On 2.11 Problem, which is still significant.  I started a type class for entity bodies with an eye toward a `Process[Task, ByteVector]` and `Stream[Task, Byte]` instance, and projected to be a fat type class with a lot of single-use operations.  This is where that shim would fail.

Of my ideas above, each of which I abhor, I think I find the -cats and -fs2 modules the least abhorrent, especially until those libraries go 1.0.

**Comment by tpolecat** [comment]
In this particular case, just by dumb luck, you can get around the `Either` problem by rewriting the import.

```
import cats.data.{ Xor => \/ }
import cats.data.Xor. { Left => -\/, Right => \/- }`
```

An irritating rewrite will be `(a |@| b)(f)` which becomes `(a |@| b).map(f)` in cats. Might be easy with scala.meta, dunno.

I like `#ifdef` and rewriting better in principle than shims because it doesn't introduce any irritants to the end user. I did a shimmy version of atto that people say works fine but it does introduce another layer of machinery that's purely for *my* benefit as a maintainer.

I'm really glad we finally got out the machetes. At least we're in motion now.

**Comment by LeifW** [comment]
@tpolecat Are you going to put all those glowing endorsements on your testimonials page?
I like `#ifdef`s because it seems like the Haskell thing to do. :|

**Comment by rossabaker** [comment]
* Yaxy solutions: all burden on maintainers
* Shimmy solutions: shared burden between maintainers and users.
* Support module solutions: shared burden on maintainers and users of the second-class libraries.

It Sucks For Someone™

**Comment by djspiewak** [comment]
We should roll our own, third functional abstraction layer.  That would seem to be the best way to make it suck for EVERYONE: users, maintainers, and random bystanders.

**Comment by rossabaker** [comment]
Suppose we stick with the scalaz/scalaz-stream foundation for now, but provide an http4s-cats and http4s-fs2 for the early adopters.  What sucks for users?

*`EntityEncoder` doesn't suck.
  * `Ok(fs2Task)`
  * `Ok(fs2Stream)`

* We can provide a `scalaz.concurrent.Task ~> fs2.Task` and back, but more sugar is needed:
  * `EntityDecoder` needs syntactic help to get to an fs2 Task:
    * `req.as[String].transform[fs2.Task]` ?
    * `req.asFs2[String]: fs2.Task` ?
    * `req.attemptAsFs2[String]: XorT[fs2.Task, MessageFailure, String]` ?
  * `Client` might merit a whole new wrapper to avoid explicit round trips like:
    * `client.fetch(req)(logic(_).transform[scalaz.concurrent.Task]).transform[fs2.Task]`

* We can provide Cats instances for any of our custom data types.  That doesn't suck.

* Web sockets kind of suck: requires a stream adapter on both source and sink.

* We can provide Cats instances for any Scalaz types that peak through (Kleisli, EitherT, \/), though there is nothing http4s about this.  Also, most uses will be direct, and in a Cats world, those suck.

(The situation is similar if cats-fs2 is the primary stack, and scalaz/scalaz-stream gets the secondary support.)

Seems adequate for a proof of concept or light use, and suffocating beyond that.

**Comment by rossabaker** [comment]
I started to sketch out a [two-branch solution](https://github.com/http4s/http4s/tree/topic/two-branch-solution).

Pros:
* No preprocessor pain.  Everything is Real Scala™
* Native types on both stacks.  No custom third monad.  (But see below about aliasing).
* No extra imports for end users.
* Most translations are isolated to compat package, greatly reducing branch divergence.

Cons:
* One branch is "secondary": the aliases will be more familiar to scalaz users than cats users.
* Accordingly, compat module is larger for secondary branch to match the primary.
* Neither branch is precisely how it would look if we supported just one stack.
* Accordingly, this would be unwound if we dropped a stack in the future.

**Comment by rossabaker** [comment]
wip/cats-fs2 is a more complete version of the above.  Rough proposal:

Source management:
* scalaz-stream is the master branch.
* cats-fs2 is maintained on a separate, actively synchronized branch.
* If most maintainers prefer to work in cats/fs2, we can and should flip these branches.

Rules of the road on the master branch:
* scalaz imports are to be kept together in their own stanza to minimize merge conflicts.  I optimized imports with IntelliJ and put all the scalaz imports at the top.
* Prefer wild card imports of scalaz types to reduce the frequency these imports change.
* Functions and instances consolidated into private `org.http4s.compat` object.  (Alternatively, embrace `import scalaz._, Scalaz._`)   Don't shy away from embracing scalaz functionality, but strive to minimize its import count.
* Aside from import hygiene, most contributors can largely ignore the ports.
* Please be considerate if someone asks for a code reorganization that aids the port branch.  Remember, that port branch is increasing the size of our community. 

Rules of the road on the port branch:
* cats and fs2 imports replace the scalaz stanza.
* Rename to match the scalaz names where necessary, to keep the diffs in the imports (e.g. `import cats.data.{Xor => \/, _}, fs2.{Stream => Process, _}`
* Extend the compat object as necessary to mimic scalaz/scalaz-stream functionality with syntax.  This may include companion objects to cats types (e.g., `Show.showA`).
* The goal is to minimize the diff with the master branch.
* New features should be added to the master branch, to benefit the whole community.
* This branch is only be as successful as maintainers who step up and keep new features compiling.  If you asked for it, help wanted.
* This branch can wither without any adverse effect on master.

What it all means:
* Users will get http4s on the stack they want, as long as they want scalaz-stream or cats-fs2-Task.  All types will appear "native."
* Contibutors on master are burdened with keeping a subset of imports clean and not much else.
* Those who want the port do the heaviest lifting, but much of this lift is initial.

**Comment by tpolecat** [comment]
What are your thoughts re: scalaz+fs2 (i.e., a third branch)?

**Comment by rossabaker** [comment]
I am willing to work on scalaz-stream for work and its production track record, and I am willing to work on cats+fs2 for fun and my belief that they are the future.  scalaz+fs2 is a stack I wouldn't care to put ongoing effort into, but would be happy to help publish if a contributor steps up.

**Comment by mdedetrich** [comment]
I would be willing to look into a cats + monix combo, as that is what we are likely to use at work

**Comment by refried** [comment]
Thanks for doing this tough work.

**Comment by rossabaker** [comment]
I appreciate everyone's thoughtful feedback on this.  After long discussions and many experiments, it's time to decide.

On the topic/cats branch, I have core compiling atop cats and fs2.  This is in my opinion the stack of the future, the preference of most contributors, and an increasing share of the community.  I want to continue this work and make this the master branch when it's ready.

I do not want to abruptly abandon our early adopters on scalaz-stream.  I will maintain and publish a scalaz-stream branch off master for as long as I am invested in this stack at work, which projects to be for several months.  This branch will reach EOL at whatever point no contributor is willing to keep it alive.

**Comment by Daenyth** [comment]
What's the current lay of the land? Is the future going to be master = cats+fs2 and there's going to be a fork for scalaz/scalaz-streams?

**Comment by rossabaker** [comment]
Yes, cats/fs2 is the future master, as soon as it's ready.

I use scalaz/scalaz-stream at work now.  I expect to maintain that branch for at least as long as my work is on scalaz/scalaz-stream.

**Comment by agustafson** [comment]
Hi, just wondering what's still outstanding for this? Is any help needed with the client work? I've finally managed to get my head around the fs2 tcp client code so might be able to help out with that @rossabaker

**Comment by rossabaker** [comment]
#661 has a checklist.  Things that aren't done are commented `TODO fs2 port`.  If you'd like to take any of those on, please leave a comment there on the ticket, and we'd love to have the help!

**Comment by agustafson** [comment]
Great ta @rossabaker, it just seemed that somebody may have been looking at the client code, so didn't want to step on their toes. I'll comment on #661 with details and see if I can find some time to make some changes.

**Comment by rossabaker** [comment]
Before anyone touches the client again, let's see if we can get #794 in for a landing.  I'll take a look right now.

**Comment by agustafson** [comment]
Sounds good ta @rossabaker

**Comment by rossabaker** [comment]
Work is now being coordinated on [the fs2-cats port project](https://github.com/http4s/http4s/projects/3).

**Comment by rossabaker** [comment]
Tentatively, we're keeping master on scalaz-stream while release-0.15.x is actively maintained to preserve the release-0.15.x -> master -> cats flow.  I anticipate we'll flip the branches after 0.16 is released to get cats front and center.

We're publishing mostly complete snapshots on cats+fs2 and have documented the gaps.  The work continues, but I think this ticket ends here.
